File: scripts/process-docs.ts
#!/usr/bin/env tsx
/**
 * process-docs.ts
 * Procesa los PDFs de mechani_ai/, extrae texto, genera embeddings (Google AI)
 * y los guarda en Supabase knowledge_base.
 *
 * Uso: npx tsx scripts/process-docs.ts [--reset]
 *   --reset: borra todos los embeddings existentes antes de procesar
 */

import fs from 'fs';
import path from 'path';

// Cargar .env.local manualmente
const envFile = path.resolve('.env.local');
if (fs.existsSync(envFile)) {
  for (const line of fs.readFileSync(envFile, 'utf8').split('\n')) {
    const [key, ...rest] = line.split('=');
    if (key && rest.length) process.env[key.trim()] = rest.join('=').trim();
  }
}
import { execSync } from 'child_process';
import { createClient } from '@supabase/supabase-js';
import { GoogleGenerativeAI } from '@google/generative-ai';

// ─── Config ───────────────────────────────────────────────────────────────────
const DOCS_DIR = path.resolve('./mechani_ai');
const CHUNK_SIZE = 800;       // caracteres por chunk
const CHUNK_OVERLAP = 100;    // solapamiento entre chunks
const BATCH_SIZE = 50;        // chunks por insert batch
const EMBED_DELAY_MS = 200;   // delay entre llamadas a embedding API (rate limit)

// Variables de entorno (requiere .env.local o env seteados)
const supabaseUrl = process.env.VITE_SUPABASE_URL || process.env.SUPABASE_URL || '';
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.VITE_SUPABASE_ANON_KEY || '';
// Preferimos GOOGLE_AI_API_KEY_INGEST (cuota separada para no tirar prod abajo).
const googleApiKey = process.env.GOOGLE_AI_API_KEY_INGEST || process.env.GOOGLE_AI_API_KEY || '';

if (!supabaseUrl || !supabaseKey || !googleApiKey) {
  console.error('❌ Faltan variables de entorno. Verificá VITE_SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY y GOOGLE_AI_API_KEY');
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseKey);
const genAI = new GoogleGenerativeAI(googleApiKey);
const embeddingModel = genAI.getGenerativeModel({ model: 'gemini-embedding-001' });

// ─── Helpers ──────────────────────────────────────────────────────────────────

function chunkText(text: string): string[] {
  const chunks: string[] = [];
  let start = 0;
  while (start < text.length) {
    const end = Math.min(start + CHUNK_SIZE, text.length);
    const chunk = text.slice(start, end).trim();
    if (chunk.length > 50) chunks.push(chunk);
    if (end >= text.length) break;
    start = end - CHUNK_OVERLAP;
  }
  return chunks;
}

async function getEmbedding(text: string): Promise<number[]> {
  const result = await embeddingModel.embedContent({ content: { parts: [{ text }] }, outputDimensionality: 768 } as any);
  return result.embedding.values;
}

function extractMetaFromPath(filePath: string): Record<string, string> {
  const relative = path.relative(DOCS_DIR, filePath);
  const parts = relative.split(path.sep);
  const filename = parts[parts.length - 1].replace('.pdf', '');

  // Inferir marca del path
  const brandKeywords = ['RENAULT', 'PEUGEOT', 'VW', 'VOLKSWAGEN', 'FORD', 'FIAT', 'GM', 'CHEVROLET', 'HYUNDAI', 'CITROEN', 'TOYOTA'];
  const pathUpper = relative.toUpperCase();
  const marca = brandKeywords.find(b => pathUpper.includes(b)) || 'GENERAL';

  return {
    filename,
    path: relative,
    marca,
    folder: parts.slice(0, -1).join('/'),
  };
}

function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function getAllPdfs(dir: string): Promise<string[]> {
  const results: string[] = [];
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      results.push(...await getAllPdfs(full));
    } else if (entry.name.toLowerCase().endsWith('.pdf')) {
      results.push(full);
    }
  }
  return results;
}

// ─── Main ─────────────────────────────────────────────────────────────────────

async function main() {
  const shouldReset = process.argv.includes('--reset');

  if (shouldReset) {
    console.log('🗑️  Borrando embeddings existentes...');
    await supabase.from('knowledge_base').delete().neq('id', '00000000-0000-0000-0000-000000000000');
    console.log('✅ Tabla limpia');
  }

  // Obtener archivos ya procesados para no duplicar
  const { data: existing } = await supabase
    .from('knowledge_base')
    .select('metadata')
    .not('metadata->>path', 'is', null);

  const processedPaths = new Set(
    (existing || []).map(r => r.metadata?.path).filter(Boolean)
  );
  console.log(`📋 Ya procesados: ${processedPaths.size} archivos únicos`);

  const allPdfs = await getAllPdfs(DOCS_DIR);
  console.log(`📁 PDFs encontrados: ${allPdfs.length}`);

  let totalChunks = 0;
  let processedFiles = 0;
  let skippedFiles = 0;
  let errorFiles = 0;

  for (const pdfPath of allPdfs) {
    const relative = path.relative(DOCS_DIR, pdfPath);

    if (processedPaths.has(relative)) {
      skippedFiles++;
      continue;
    }

    console.log(`\n📄 [${processedFiles + 1}/${allPdfs.length}] ${relative}`);

    try {
      const text = execSync(`pdftotext "${pdfPath}" -`, { maxBuffer: 10 * 1024 * 1024, timeout: 15000 }).toString().trim();

      if (!text || text.length < 100) {
        console.log(`  ⚠️  Texto insuficiente (${text?.length || 0} chars) — saltando`);
        skippedFiles++;
        continue;
      }

      console.log(`  ✏️  Texto extraído: ${text.length} chars`);

      const chunks = chunkText(text);
      console.log(`  🔪 Chunks: ${chunks.length}`);

      const meta = extractMetaFromPath(pdfPath);
      let savedChunks = 0;

      // Generar embedding e insertar de inmediato (sin acumular en RAM)
      for (let i = 0; i < chunks.length; i++) {
        const chunk = chunks[i];
        await sleep(EMBED_DELAY_MS);
        const embedding = await getEmbedding(chunk);
        const { error } = await supabase.from('knowledge_base').insert({
          content: chunk,
          metadata: { ...meta, chunk: String(i) },
          embedding,
        });
        if (error) throw error;
        savedChunks++;
        process.stdout.write(`\r  🧠 ${savedChunks}/${chunks.length} chunks`);
      }
      console.log();

      totalChunks += savedChunks;
      processedFiles++;
      console.log(`  ✅ Guardados ${savedChunks} chunks`);

    } catch (err) {
      console.error(`  ❌ Error:`, (err as Error).message);
      errorFiles++;
    }
  }

  console.log('\n─────────────────────────────────────');
  console.log(`✅ Procesados: ${processedFiles} archivos`);
  console.log(`⏭️  Saltados:   ${skippedFiles} (ya existían o sin texto)`);
  console.log(`❌ Errores:    ${errorFiles}`);
  console.log(`📦 Total chunks insertados: ${totalChunks}`);
}

main().catch(err => {
  console.error('Fatal:', err);
  process.exit(1);
});

File: scripts/process-docs.py
#!/usr/bin/env python3
"""
process-docs.py
Procesa los PDFs de mechani_ai/, extrae texto con pdftotext,
genera embeddings con Google AI y los guarda en Supabase.
"""

import os, sys, json, time, subprocess, urllib.request, urllib.parse
from pathlib import Path

# ─── Cargar .env.local ────────────────────────────────────────────────────────
env_file = Path('.env.local')
if env_file.exists():
    for line in env_file.read_text().splitlines():
        if '=' in line and not line.startswith('#'):
            k, _, v = line.partition('=')
            os.environ.setdefault(k.strip(), v.strip())

SUPABASE_URL  = os.environ.get('VITE_SUPABASE_URL') or os.environ.get('SUPABASE_URL', '')
SUPABASE_KEY  = os.environ.get('SUPABASE_SERVICE_ROLE_KEY', '')
GOOGLE_KEY    = os.environ.get('GOOGLE_AI_API_KEY', '')

if not all([SUPABASE_URL, SUPABASE_KEY, GOOGLE_KEY]):
    print('❌ Faltan variables de entorno: VITE_SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY, GOOGLE_AI_API_KEY')
    sys.exit(1)

DOCS_DIR     = Path('./mechani_ai')
CHUNK_SIZE   = 800
CHUNK_OVERLAP = 100
EMBED_DELAY  = 0.2
RESET        = '--reset' in sys.argv

# ─── Helpers ──────────────────────────────────────────────────────────────────

def http_post(url, data, headers):
    body = json.dumps(data).encode()
    req = urllib.request.Request(url, data=body, headers={**headers, 'Content-Type': 'application/json'})
    with urllib.request.urlopen(req, timeout=30) as r:
        return json.loads(r.read())

def get_embedding(text):
    url = f'https://generativelanguage.googleapis.com/v1beta/models/gemini-embedding-001:embedContent?key={GOOGLE_KEY}'
    resp = http_post(url, {'content': {'parts': [{'text': text}]}, 'outputDimensionality': 768}, {})
    return resp['embedding']['values']

def supabase_insert(row):
    url = f'{SUPABASE_URL}/rest/v1/knowledge_base'
    headers = {'apikey': SUPABASE_KEY, 'Authorization': f'Bearer {SUPABASE_KEY}', 'Prefer': 'return=minimal'}
    http_post(url, row, headers)

def supabase_get_processed():
    url = f'{SUPABASE_URL}/rest/v1/knowledge_base?select=metadata&limit=10000'
    req = urllib.request.Request(url, headers={'apikey': SUPABASE_KEY, 'Authorization': f'Bearer {SUPABASE_KEY}'})
    with urllib.request.urlopen(req, timeout=30) as r:
        rows = json.loads(r.read())
    return {r['metadata'].get('path') for r in rows if r.get('metadata', {}).get('path')}

def supabase_reset():
    url = f'{SUPABASE_URL}/rest/v1/knowledge_base?id=neq.00000000-0000-0000-0000-000000000000'
    req = urllib.request.Request(url, method='DELETE', headers={'apikey': SUPABASE_KEY, 'Authorization': f'Bearer {SUPABASE_KEY}'})
    with urllib.request.urlopen(req, timeout=30): pass

def chunk_text(text):
    chunks, start = [], 0
    while start < len(text):
        end = min(start + CHUNK_SIZE, len(text))
        chunk = text[start:end].strip()
        if len(chunk) > 50:
            chunks.append(chunk)
        if end >= len(text):
            break
        start = end - CHUNK_OVERLAP
    return chunks

def extract_text(pdf_path):
    try:
        result = subprocess.run(
            ['pdftotext', str(pdf_path), '-'],
            capture_output=True, text=True, timeout=15
        )
        return result.stdout.strip()
    except subprocess.TimeoutExpired:
        raise Exception(f'pdftotext timeout (>15s) — PDF probablemente corrupto o muy complejo')

def get_marca(relative):
    brands = ['RENAULT', 'PEUGEOT', 'VW', 'VOLKSWAGEN', 'FORD', 'FIAT', 'GM', 'CHEVROLET', 'HYUNDAI', 'CITROEN', 'TOYOTA']
    upper = relative.upper()
    return next((b for b in brands if b in upper), 'GENERAL')

# ─── Main ─────────────────────────────────────────────────────────────────────

if RESET:
    print('🗑️  Borrando embeddings existentes...')
    supabase_reset()
    print('✅ Tabla limpia')

processed_paths = supabase_get_processed()
print(f'📋 Ya procesados: {len(processed_paths)} archivos únicos')

all_pdfs = sorted(DOCS_DIR.rglob('*.pdf'))
print(f'📁 PDFs encontrados: {len(all_pdfs)}')

total_chunks = processed_files = skipped_files = error_files = 0

for i, pdf_path in enumerate(all_pdfs):
    relative = str(pdf_path.relative_to(DOCS_DIR))

    if relative in processed_paths:
        skipped_files += 1
        continue

    print(f'\n📄 [{i+1}/{len(all_pdfs)}] {relative}')

    try:
        text = extract_text(pdf_path)
        if len(text) < 100:
            print(f'  ⚠️  Texto insuficiente ({len(text)} chars) — saltando')
            skipped_files += 1
            continue

        print(f'  ✏️  Texto extraído: {len(text)} chars')
        chunks = chunk_text(text)
        print(f'  🔪 Chunks: {len(chunks)}')

        parts = Path(relative).parts
        meta = {
            'filename': pdf_path.stem,
            'path': relative,
            'marca': get_marca(relative),
            'folder': '/'.join(parts[:-1]),
        }

        for j, chunk in enumerate(chunks):
            time.sleep(EMBED_DELAY)
            embedding = get_embedding(chunk)
            supabase_insert({'content': chunk, 'metadata': {**meta, 'chunk': str(j)}, 'embedding': embedding})
            print(f'\r  🧠 {j+1}/{len(chunks)} chunks', end='', flush=True)

        print()
        total_chunks += len(chunks)
        processed_files += 1
        print(f'  ✅ Guardados {len(chunks)} chunks')

    except Exception as e:
        print(f'  ❌ Error: {e}')
        error_files += 1

print('\n─────────────────────────────────────')
print(f'✅ Procesados: {processed_files} archivos')
print(f'⏭️  Saltados:   {skipped_files} (ya existían o sin texto)')
print(f'❌ Errores:    {error_files}')
print(f'📦 Total chunks insertados: {total_chunks}')
